### DISCUSSION TITLE
Newbie Question: how to clip a curved line between 2 given points

### DISCUSSION THREAD

**Discussion by bancroftway** [NetTopologySuite/NetTopologySuite#600] [Q&A]
I have a line (LineString) that passes through the red triangle and the black X, and looks like this:
![image](https://user-images.githubusercontent.com/13262438/157552450-72edbd72-86cd-412a-a702-72455a0b09cf.png)

I want to clip the line to only the segment between the red triangle and black X. 

I used this code:

```
LengthLocationMap llm = new LengthLocationMap(ls);   //where ls is my LineString
LengthIndexedLine lil = new LengthIndexedLine(ls); //where ls is my LineString
ExtractLineByLocation ell = new ExtractLineByLocation(ls); //where ls is my LineString        

LinearLocation ll1 = llm.GetLocation(lil.Project(redTriangle.Coordinate)); 
   
LinearLocation ll2 = llm.GetLocation(lil.Project(blackX.Coordinate), true);

var ls2 = ell.Extract(ll1, ll2); //This is the clipped line (final result)
```

The above code produced the result below, which is not what I had in mind. Could you please suggest a fix in the code above, such that I can keep just the line segment between the red triangle and black X? Thanks

![image](https://user-images.githubusercontent.com/13262438/157553826-0aa9a4ea-e0e7-4e32-b491-e2d01feb9328.png)

**Comment by FObermaier** [comment]
Which version of NTS are you using? The following code seem to return a correct result.
```C#
const string wktLine = "LINESTRING (0 10, 5 12, 15 8, 14 7.4, 5 2)";
const string wktTriangle = "POINT (4.99 12.01)";
const string wktCross = "POINT (14.01 7.39)";

var rdr = new WKTReader();
var line = rdr.Read(wktLine);
var triangle = rdr.Read(wktTriangle);
var cross = rdr.Read(wktCross);

var lil = new LengthIndexedLine(line);
double posTriangle = lil.Project(triangle.Coordinate);
double posCross = lil.Project(cross.Coordinate);

var extracted = lil.ExtractLine(posTriangle, posCross);
// LINESTRING (4.994827586206897 11.99793103448276, 5 12, 15 8, 14.00294117647059 7.4017647058823535)

var llm = new LengthLocationMap(line);
var locTriangle = llm.GetLocation(posTriangle);
var locCross = llm.GetLocation(posCross);

var ell = new ExtractLineByLocation(line);
var extracted2 = ell.Extract(locTriangle, locCross);
// LINESTRING (4.994827586206897 11.99793103448276, 5 12, 15 8, 14.00294117647059 7.4017647058823535)
```

If you can't verify that, we'd need to see the WKTs of the involved geometries.

**Comment by bancroftway** [comment]
@FObermaier I am using latest stable NTS 2.4.0 inside a .Net 6 project.

I have reproduced my issue in a Console app in github repo [here](https://github.com/bancroftway/NetTopologyIssues/tree/Issue597).

I have verified that the final extracted line looks like this, which is not what I was expecting:

![image](https://user-images.githubusercontent.com/13262438/157692473-f7072265-f60d-425d-9075-24eaa286b87c.png)

**Comment by FObermaier** [comment]
Where does that `LineString` come from? This is an image of -I'd say 1%o- of the the `LineString`:
![grafik](https://user-images.githubusercontent.com/2065346/157827305-5c0cde9c-2df0-4a06-bf0b-80689659084d.png)
```
LINESTRING (2317753.377998376 796346.5849989518, 2317753.377998376 796346.5949989515, 2317753.377998376 796346.6049989513, 
2317753.377998376 796346.6149989511, 2317753.377998376 796346.6249989509, 2317753.377998376 796346.6349989506, 
2317753.377998376 796346.6449989504, 2317753.377998376 796346.6549989502, 2317753.377998376 796346.66499895, 
2317753.3879983756 796346.5849989518, 2317753.3879983756 796346.5949989515, 2317753.3879983756 796346.6049989513, 
2317753.3879983756 796346.6149989511, 2317753.3879983756 796346.6249989509, 2317753.3879983756 796346.6349989506, 
2317753.3879983756 796346.6449989504, 2317753.3879983756 796346.6549989502, 2317753.3879983756 796346.66499895, 
2317753.3979983754 796346.5849989518, 2317753.3979983754 796346.5949989515, 2317753.3979983754 796346.6049989513, 
2317753.3979983754 796346.6149989511, 2317753.3979983754 796346.6249989509, 2317753.3979983754 796346.6349989506, 
2317753.3979983754 796346.6449989504, 2317753.3979983754 796346.6549989502, 2317753.3979983754 796346.66499895, 
2317753.407998375 796346.5849989518, 2317753.407998375 796346.5949989515, 2317753.407998375 796346.6049989513, 
2317753.407998375 796346.6149989511, 2317753.407998375 796346.6249989509, 2317753.407998375 796346.6349989506, 
2317753.407998375 796346.6449989504, 2317753.407998375 796346.6549989502, 2317753.407998375 796346.66499895, 
2317753.417998375 796346.5949989515, 2317753.417998375 796346.6049989513, 2317753.417998375 796346.6149989511, 
2317753.417998375 796346.6249989509, 2317753.417998375 796346.6349989506, 2317753.417998375 796346.6449989504, 
2317753.417998375 796346.6549989502, 2317753.417998375 796346.66499895, 2317753.4279983747 796346.5949989515, 
2317753.4279983747 796346.6049989513, 2317753.4279983747 796346.6149989511, 2317753.4279983747 796346.6249989509, 
2317753.4279983747 796346.6349989506, 2317753.4279983747 796346.6449989504, 2317753.4279983747 796346.6549989502, 
2317753.4279983747 796346.66499895, 2317753.4379983745 796346.5949989515, 2317753.4379983745 796346.6049989513, 
2317753.4379983745 796346.6149989511)
```
I assume the rest is similar to that, with a little back and forth. If you want to get the algorithm working correctly, you need to identify the extraction points exactly not approximately.

My suggestion: Get that LineString fixed.

**Comment by bancroftway** [comment]
@FObermaier the linestring was obtained by first calculating a quantity representing air flow "head" at fixed intervals on a grid, and then selecting a particular value from this mesh which resulted in the line string above,...that is why there is a the back-and-forth. Is there a way you can suggest to take this data and create a better line string out of it? Is Simplifier an option?

**Comment by FObermaier** [comment]
I'm afraid I don't quite comprehend what you are doing, it doesn't seem to be my domain. Are you trying to build contour lines?

**Reply by bancroftway** [reply]
@FObermaier  Yes you can think of these as contour lines. I start at the bottom left origin of the field and I take a fixed step in both x and y directions, and at each point, I calculate a quantity Si, using a formula. Having calculated a mesh of Si values, I select all points on the mesh whose Si value equals to the Si value at the black X point..this should result in a "streamline". However, because of the way I am computing this mesh the line is not as smooth as desired and is jagged as you saw above. 

I need to clip the line between red triangle and black X.

This problem is admittedly from a different domain...but I was hoping to use NTS to help me do the clipping.

Can you think of a way, such as using Simplifiers, to "fix" the above linestring?
